bip-0340.mediawiki

```diff
@@ -155,7 +156,7 @@ The algorithm ''Sign(sk, m)'' is defined as:
 * Fail if ''d' = 0'' or ''d' &ge; n''
 * Let ''P = d'⋅G''
 * Let ''d = d' '' if ''has_even_y(P)'', otherwise let ''d = n - d' ''.
-* Let ''rand = hash<sub>BIP340/nonce</sub>(bytes(d) || m)''.
+* Let ''rand = hash<sub>BIP340/nonce</sub>(bytes(d) || m || extbytes(P))''<ref>Including the public key as input to the nonce hash helps ensure the robustness of the signing algorithm, preventing leakage of the secret key if the calculation of the public key ''P'' is performed incorrectly or maliciously, eg due to being left to the caller for performance reasons.</ref>.
```

Comment:
Nit: "." after "eg"

Maybe link to https://moderncrypto.org/mail-archive/curves/2020/001012.html (after the site's certificate is updated...).

Reply:
A question I have is whether we should then also use `bytes(d')` (= `sk`) rather than `bytes(d)`, if the signature is going to be different depending on the key's sign.

Reply:
@gmaxwell suggested putting the privkey and pubkey first, so that for a fixed key, the midstate can be precomputed.

Reply:
Done

Reply:
> A question I have is whether we should then also use `bytes(d')` (= `sk`) rather than `bytes(d)`, if the signature is going to be different depending on the key's sign.

I don't think it matters in the end but from my point of view, `d` and `n-d` are simply different representations of the same secret key. Or in other words, there is only one real secret work (the normalized one), and the only reason that sign accepts both is that we want to be compatible with existing methods to generate secret keys.

Moreover, we have a "pk <- keygen(sk)" API (instead of the more common "(sk,pk) <- keygen()", see also djb's reply on the mailing list. To be honest, this now makes me wonder what we can do about this. If we think that implementations should have an secret key object that carries both d and pk, and we want to do it like that in our own implementation, shouldn't we write the BIP like this in the first place? At the moment our signing API in the BIP is okay because it does not precompute public keys. But its inefficiency encourages implementors to precompute pk (this is the entire reason for including pk in the derivation), and then we perhaps should hold their hand for the right method to store secret keys. 

This would change the API to 
(sk, pk) <- keygen(randomness), but in the implementation sk will include a copy of pk 
sig <- sign(sk, m) , and this is fast by default
0/1 <- verify(pk, m), no change here

The advantage here is that this is safe and fast, so noone needs to be clever and tweak it.

Reply:
As mentioned in the original issue we do expect implementors to take theprecomputed pubkey shortcut and therefore we need to explain how to do this inan unmistakable way. This PR's advice that "signers must ensure the public keyis correctly calculated and not taken from untrusted sources." is easy to miss. I do think this PR would be okay, but I prefer speccing it out and@real-or-random's suggestion looks quite clean actually. I also don't see adownside in the ergonomics of libsecp. We can offer a convenience `sign`function that takes a char array and does keygen and sign in one step.

Reply:
So, realized that we may not actually need the negation flag / extpubkey in the derivation function.

Our realization earlier was that "all information that goes into the signing algorithm must go into the nonce derivation as well".

We currently have a signing algorithm that is a function of d', and P=d'G. However, this input is transformed into either (d', P) or (-d', -P). The actual signing algorithm can be expressed in function of just this transformed data, which implies that as long as it is d that goes into the nonce and not d' or sk, it should not be necessary to include P's even/oddness.

This allows for better alignment, but if we worry that someone might feed sk or d' into the nonce derivation instead, maybe it's better to prescribe (d', extbytes(P)). Of course, they may equally make a mistake in encoding P's sign, so perhaps this does not add anything.

Reply:
I've update the PR to define `sk = KeyGen(rng32bytes)` to be 64 bytes and `pk = PubKey(sk) = sk[32:64]` and corresponding simplification to `Sign` which looks pretty good to me fwiw

Reply:
Just to clarify the using `d` instead of `d'` is only an issue if you're using precomputed pubkeys, but somehow you have the wrong negation flag (and then produce an invalid sig).
I don't think this is easy to mess up by implementors if we have test vectors that catch this and perhaps add a footnote with a rationale.

But a better solution is to go with @real-or-random's suggestion and negate the secret key right at keygen, because then there is no negation flag to mess up.